internal/server/backgroundjobs.go

```diff
@@ -0,0 +1,77 @@
+package server
+
+import (
+	"context"
+	"time"
+
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/jobs"
+)
+
+// BackgroundJobFunc is the interface for implementing a new background job.
+//
+// currentTime is the time the job was invoked at, and should be used for segmenting records into processable chunks.
+// Jobs may want to ignore records past the currentTime, as those will be handled in a future job run.
+//
+// lastRunAt is the time that the background job was last invoked. This value will only be updated if the function returns a non-error
+//
+// errors will be logged but will not cause the app to crash.
+//
+// panics will be caught and logged
+//
+// jobs should gracefully exit if their context quits, eg ctx.Done() or ctx.Err()
+type BackgroundJobFunc func(ctx context.Context, tx data.WriteTxn, lastRunAt, currentTime time.Time) error
+
+func (s *Server) SetupBackgroundJobs() {
+	s.registerJob(jobs.RemoveOldDeviceFlowRequests, 10*time.Minute)
+	s.registerJob(jobs.RemoveExpiredAccessKeys, 1*time.Minute)
+	s.registerJob(jobs.RemoveExpiredPasswordResetTokens, 15*time.Minute)
+}
+
+func (s *Server) registerJob(job BackgroundJobFunc, every time.Duration) {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.routines = append(s.routines, routine{
+		run:  jobWrapper(ctx, s.db, job, every),
+		stop: cancel,
+	})
+}
+
+func jobWrapper(ctx context.Context, tx data.WriteTxn, job BackgroundJobFunc, every time.Duration) func() error {
+	return func() error { // jobs shouldn't really return errors, we just do this to be compatible with the "routine" struct.
+		t := time.NewTicker(every)
+		lastRunAt := time.Time{}
+
+		jobWithRescue := func() {
+			if ctx.Err() != nil {
+				return
+			}
+			defer func() {
+				if err := recover(); err != nil {
+					logging.Errorf("background job %s panic: %s", getFuncName(job), err)
+				}
+			}()
+
+			startAt := time.Now().UTC()
+			logging.Debugf("background job %s starting", getFuncName(job))
+
+			// TODO: consider attaching the context to the tx so long-running executions don't hold up app shutdown
+			err := job(ctx, tx, lastRunAt, startAt)
```

Comment:
I think Bruce commented on this as well, but it seems like these two times are not used.

`startAt` does not seem necessary. A job can call `time.Now` itself to get the current time.

`lastRunAt` I could see being useful, but again a job should be able to track that for itself by using a struct or closure for the job function. 

I think we benefit from keeping this interface small, so I think we should remove these two args from the interface.

Reply:
a job could do it itself, but this seems like a fine thing to give to the job runner. Why should every job that wants it reimplement time tracking logic? Can always remove it later if no jobs end up using it.

Reply:
Right now no jobs need time tracking logic. This is technical debt that will need to be cleaned up, why introduce it in the first place?

Reply:
Personally I would leave a lot more up to the individual job. There's no real benefit in abstracting things like a for loop. Better to repeat a little bit to make the code more obvious. We could have just a function that accepts a time.Duration and does the select, and leave the rest up to the job. That's roughly the pattern I used in https://github.com/infrahq/infra/blob/main/internal/repeat/waiter.go and it allows the code to be used in a few different ways without having a bunch of parameters.

I think this PR strikes a pretty nice balance. It don't think it hurts to combine the for loop and the select and recovery all together, that seems quite reasonable. But the time tracking is not necessary and distracts from the parts that are being used.